ensembleperturbation/perturbation/atcf.py

```diff
@@ -601,56 +727,198 @@ def __init__(self):
             unit=units.nautical_mile,
         )
 
+    def find_GAHM_parameters(self, B, Ro_inv) -> Quantity:
+        """ 
+        find Bg and phi for GAHM model iteratively
+        with given B and Rossby number (inverse)
+        """
+        Bg = B
+        Bm = 1e6
+        while any(abs(Bm - Bg) > 1e-3):
+            Bm = Bg
+            phi = 1 + Ro_inv / (Bg * (1 + Ro_inv))
+            Bg = B * (1 + Ro_inv) * numpy.exp(phi - 1) / phi
+        return Bg, phi
+
+    def find_parameter_from_GAHM_profile(
+        self,
+        Vr,
+        Vmax,
+        f,
+        Ro_inv,
+        B=None,
+        Bg=None,
+        phi=None,
+        Rmax=None,
+        isotach_rad=None,
+        Rrat=None,
+    ) -> Quantity:
+        """ 
+        Use root finding algorithm to return a desired parameter 
+        based on the GAHM profile of velocity that matches the input Vr
+        Input B (guess), isotach_rad and Rrat to get back the actual B
+        Input Bg, phi, Rmax, and Rrat to get back isotach_rad 
+        """
+        if B is not None:
+            # initial guesses when trying to find Bg, phi (and new B)
+            Bg, phi = self.find_GAHM_parameters(B, Ro_inv)
+            rfo2 = 0.5 * isotach_rad * f
+            alpha = Rrat ** Bg
+        Vr_test = 1e6 * MaximumSustainedWindSpeed.unit
+        tol = 1e-2 * MaximumSustainedWindSpeed.unit
+        while any(abs(Vr_test - Vr) > tol):
+            if B is None:
+                # updates for when trying to find isotach_rad
+                isotach_rad = Rmax / Rrat
+                rfo2 = 0.5 * isotach_rad * f
+                alpha = Rrat ** Bg
+            Vr_test = (
+                numpy.sqrt(
+                    Vmax ** 2 * (1 + Ro_inv) * numpy.exp(phi * (1 - alpha)) * alpha + rfo2 ** 2
+                )
+                - rfo2
+            )
+            Vr_test[Vr_test < tol] = numpy.nan  # no solution
+            # bi-section method
+            alpha[Rrat <= 1] *= 0.5 * (1 + (Vr / Vr_test)[Rrat <= 1] ** 2)
+            alpha[Rrat > 1] *= 0.5 * (1 + (Vr_test / Vr)[Rrat > 1] ** 2)
+            if B is not None:
+                # update to and Bg, phi
+                Bg = numpy.log(alpha) / numpy.log(Rrat)
+                phi = 1 + Ro_inv / (Bg * (1 + Ro_inv))
+                phi[phi < 1] = 1
+            else:
+                # update to Rrat
+                Rrat = numpy.exp(numpy.log(alpha) / Bg)
+        if B is not None:
+            return Bg * phi / ((1 + Ro_inv) * numpy.exp(phi - 1))  # B
+        else:
+            return Rmax / Rrat  # isotach_rad
+
+    def get_isotach_properties(self, dataframe: DataFrame) -> Quantity:
```

Comment:
Used to find attributes of storm and the radial velocity of the isotach at the top of the boundary layer for this quadrant